### ISSUE TITLE
Signed VbaProject is not allowed for opening when configuring Excel to only allow signed Macros

### ISSUE THREAD

**Issue by DMW007** [EPPlusSoftware/EPPlus#194]
## Motivation and target

We only want to allow the execution of signed MS Excel macros. This should allow only self made macros and forbid third party macros which potentially could contain any malicious VBA code. To let users sign their macros easily, I'd like to develop a web application where they can upload their unsigned excel files and download them signed. I'm currently evaluating EPPlus for this purpose, so I developed a simple ASP.NET Core web app as POC.

## Problem: MS Excel doesn't seem to detect them correctly as signed
To test the signed macro file together with excels policy, I went to the _trust center_ settings > _macro settings_ and enabled _Disable all macros except digitally signed macros_. It's the highlighted entry we enabled on a test vm:

![grafik](https://user-images.githubusercontent.com/21245043/95089479-7b131f80-0724-11eb-8ea6-f505a5156dac.png)

Now I'd expect that the signed macro could be opened and the unsigned wont open. But after opening the signed file, I got the following error message:

![grafik](https://user-images.githubusercontent.com/21245043/95093889-8fa5e680-0729-11eb-9354-badd7e34bbe9.png)

English translation: _SECURITY WARNING: Macros were deactivated_

![grafik](https://user-images.githubusercontent.com/21245043/95089981-0f7d8200-0725-11eb-8bac-7697dea40c46.png)

English:

> Microsoft Office security options
> Safety Instructions - Multiple Problems
>Macros
>Macros were deactivated. Macros can contain viruses or other security risks. Only enable this content if you trust the source of this file.
>Warning: Macros are disabled because the digital signature is invalid. You can view the certificate information associated with the signature. You can also activate the macros. Click the "More Information" link for more information.

## Why is the signature invalid?
So Excel says the signature is invalid. But the certificate is valid. [According to the docs](https://github.com/EPPlusSoftware/EPPlus/wiki/VBA), I checked `cert.HasPrivateKey` which is true. Start date is in 2018 and end in 2023. I also can confirm that the cert is added successfully when opening _Developer tools_ > _Visual Basic_ > _Extras_ > _Digital signature_

![grafik](https://user-images.githubusercontent.com/21245043/95091160-65065e80-0726-11eb-9b82-d7254ccbb126.png)

Here it shows me that the VBA project is currently signed with my used certificate. When clicking on _Details_ it shows me my certificate provided by our CA. It must have something to do with EPPlus since we also tried signing the test file manually. This way it could be opened without issues, only when signed with EPPlus the above error occurs.

Noticeable is the reduced file size. A simple unsigned test file is 12KB large, while it only has 9.87KB after signing. On another test file I have 10.5KB unsigned and 9.67KB signed.

## Example code
The following is my POC code for a HTML file upload form:

```csharp
        [HttpPost]
        [IgnoreAntiforgeryToken]
        public IActionResult Index(IFormFile excelFile)
        {
            logger.LogInformation($"Uploaded file: {excelFile.FileName} ({excelFile.Length / 1024} KB)");
            var fileStream = excelFile.OpenReadStream();

            string certPath = Environment.GetEnvironmentVariable("CERT_PATH");
            string certPw = Environment.GetEnvironmentVariable("CERT_PW");

            var cert = new X509Certificate2(certPath, certPw);
            var xl = new ExcelPackage(fileStream);
            var wb = xl.Workbook;
            wb.VbaProject.Signature.Certificate = cert;
            byte[] signedFileData = xl.GetAsByteArray();

            var fileInfo = new FileInfo(excelFile.FileName);
            string newFileName = fileInfo.Name.Replace(fileInfo.Extension, $"_SIGNED{fileInfo.Extension}");
            logger.LogInformation($"Signed file {newFileName} ({signedFileData.Length / 1024} KB)");

            return File(signedFileData, "application/msexcel", newFileName);
        }
```

Tried on an ASP.NET Core 3.1 project with the following csproj file:

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="EPPlus" Version="4.5.3.3" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation" Version="3.1.8" />
  </ItemGroup>
</Project>
```

For testing purpose, I upgraded to the latest 5.3.2 version but still the same result.

Why do this occur and how can we fix it?

**Comment by JanKallman** [comment]
I haven't heard this before. 

Here is some code that works in 5.4 (soon to be released) with a selfsigned certificate ...
Powershell to create selfsigned cert...
```ps
New-SelfSignedCertificate -DnsName test@yourcompany.com -Type CodeSigning -CertStoreLocation cert:\CurrentUser\My
```

C# Test
```C#
            using (var p = OpenPackage("vbaSign.xlsm"))
            {
                p.Workbook.CreateVBAProject();
                var ws=p.Workbook.Worksheets.Add("Test");
                ws.Drawings.AddShape("Drawing1", eShapeStyle.Rect);
                
                //Now add some code to update the text of the shape...
                var sb = new StringBuilder();

                sb.AppendLine("Private Sub Workbook_Open()");
                sb.AppendLine("    [Test].Shapes(\"Drawing1\").TextEffect.Text = \"This text is set from VBA!\"");
                sb.AppendLine("End Sub");
                p.Workbook.CodeModule.Code = sb.ToString();

                //Optionally, Sign the code with your company certificate.
                X509Store store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
                store.Open(OpenFlags.ReadOnly);
                p.Workbook.VbaProject.Signature.Certificate = store.Certificates[5];


                SaveAndCleanup(p);
            }
```

**Comment by DMW007** [comment]
This seems to use the Windows cert store. I'm developing on ASP.NET Core to deploy the application on Linux, where it should run in Docker containers. Fetching the cert from a file works platform independent. Could also be easily deployed on Kubernetes where we just create a secret for the cert and mount it in the desired fs location. 

For me it's also the first time hearing about this: Before we enabled _Disable all macros except digitally signed macros._ I assumed that EPPlus works correctly because MS Excel 2016 shows in the Visual Basic developer tools that the macros were signed with out cert, as shown in the screenshots of the ticket.

What happens to you when you sign an excel macro file using your code, configure excel to only allow signed macro and then try to open the signed file?

The issue could be localized to EPPlus because manually signing works. Also tried [Aspose.Cells](https://docs.aspose.com/cells/net/). I fund a bug there when the signing certificate is passed as `X509Certificate2`. It only works with an overload of the class that accepts the cert as byte array as described here: https://forum.aspose.com/t/sign-vba-project-on-net-core-api/208463/11?u=th3dan

After passing the cert as byte array, it works. Even when only signed macros were allowed in the excel settings. The signed files by Aspose.Cells are a bit larger. I'm not familiar with the MS excel file format, but for me it looks like that Aspose.Cells add the certifiate on another place, allowing excel to detect the macro as signed when opening. This step seems missing in EPPlus, so that the visual basic dev tools shows the macro as signed, but excel wont detect it on opening with the macro policy.

**Comment by JanKallman** [comment]
Interesting. EPPlus uses the Windows cert store, so using it via Linux will currently not work.
But it should. I will have a look at it. My guess is that it's not that complicated.

Signing code can be found here... [ExcelVBASignature](https://github.com/EPPlusSoftware/EPPlus/blob/develop/src/EPPlus/Vba/ExcelVBASignature.cs)
Lots of bits and bytes.

**Comment by JanKallman** [comment]
Applied a fix for this issue. 
Try the latest update from our developer nuget feed (add it inside Visual Studio) to verify if it works for you
https://ci.appveyor.com/nuget/epplus

**Comment by JanKallman** [comment]
Fixed in 5.4.1
